src/main/java/com/flightstats/hub/dao/aws/S3AccessMonitor.java

```diff
@@ -0,0 +1,92 @@
+package com.flightstats.hub.dao.aws;
+
+import com.amazonaws.services.s3.model.DeleteObjectRequest;
+import com.amazonaws.services.s3.model.DeleteVersionRequest;
+import com.amazonaws.services.s3.model.GetObjectRequest;
+import com.amazonaws.services.s3.model.ObjectMetadata;
+import com.amazonaws.services.s3.model.PutObjectRequest;
+import com.amazonaws.services.s3.model.PutObjectResult;
+import com.flightstats.hub.model.Content;
+import com.flightstats.hub.model.ContentKey;
+import lombok.extern.slf4j.Slf4j;
+import org.joda.time.DateTime;
+
+import javax.inject.Inject;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+
+@Slf4j
+public class S3AccessMonitor {
+    private final HubS3Client hubS3Client;
+    private final S3BucketName s3BucketName;
+    private final Content content;
+    private static final String CHANNEL_NAME = "S3_HEALTH_MONITOR";
+
+    @Inject
+    public S3AccessMonitor(HubS3Client hubS3Client, S3BucketName s3BucketName) {
+        this.hubS3Client = hubS3Client;
+        this.s3BucketName = s3BucketName;
+        String data = "1";
+        DateTime insertTime = new DateTime();
+        this.content = Content.builder()
+                .withContentKey(new ContentKey(insertTime))
+                .withData(data.getBytes())
+                .withContentType("text/plain")
+                .build();
+    }
+
+    private String key() {
+        Optional<String> url = content.getContentKey().map(ContentKey::toUrl);
+        String keyPath = url.orElse("");
+        return CHANNEL_NAME + "/" + keyPath;
+    }
+
+    private CompletableFuture<PutObjectResult> waitForWrite() {
+        try {
+            ObjectMetadata metadata = S3SingleContentDao.createObjectMetadata(content, false);
+            metadata.setContentLength(content.getData().length);
+            PutObjectRequest request = new PutObjectRequest(s3BucketName.getS3BucketName(), key(), content.getStream(), metadata);
+            return CompletableFuture.supplyAsync(() -> hubS3Client.putObject(request));
+        } catch (Exception e) {
+            log.error("error writing to s3: ", e);
+            throw e;
+        }
+    }
+
+    private CompletableFuture<String> waitForRead(String versionId) {
+        try {
+            return CompletableFuture.supplyAsync(() -> {
+                hubS3Client.getObject(new GetObjectRequest(s3BucketName.getS3BucketName(), key(), versionId));
+                return versionId;
+            });
+        } catch(Exception e) {
+            log.error("error getting object from s3", e);
+            throw e;
+        }
+    }
+
+    private void cleanup(String versionId) {
+        try {
+            DeleteObjectRequest deleteObjectRequest = new DeleteObjectRequest(s3BucketName.getS3BucketName(), key());
```

Comment:
I'm not familiar with s3 versioning, but do you want to execute the delete both with and without the version or should this be an if/else?

Or maybe a 1-line comment before the if as to why it's in addition to, rather than instead of, the main delete.

Reply:
I mistook the docs on the delete version method on first pass, I thought that it only deleted the "delete marker" but it does delete the object as well, updated.